Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,218 @@
+.. SPDX-License-Identifier: GPL-2.0-only
+
+=============
+Ancillary Bus
+=============
+
+See <linux/ancillary_bus.h> for the models for ancillary_device and
+ancillary_driver.
+
+In some subsystems, the functionality of the core device (PCI/ACPI/other) may
```

Comment:
Start by saying 

Motivation and purpose of ancillary bus.
---------------------------------------------
Ancillary bus is intended to address below different use cases.
1. ... single device different subsystem ... MFD not suitable platform not suitable description with audio example
Here cut of the 'may' part. Please definitive and you already have examples.
So right away say what purpose is served.

2. A PCI function support multiple class of devices such as RDMA, netdevice, and vDPA.
In this scenario, it is desired to maintain different class drivers for each of the class.
For such PCI device, for each class, an ancillary device with unique id should be created. Matching driver binds to such ancillary device to create respective class device. This enables multiple class drivers to share same PCI function and shared resources such as PCI BAR regions, irqs and more.

3. A PCI sub function is another type of device which represents a portion of the PCI device, similar to PCI virtual function. Such device also needs matching service functionality. In this use case, for each PCI sub-function, an ancillary device is created with unique id. A driver binds to such ancillary device to make use of a PCI sub-function device. A PCI sub function based ancillary device will have one or more device attributes such as sub-function number, NUMA node information, PCI BAR resources and more.

Reply:
> Start by saying
> 
> ## Motivation and purpose of ancillary bus.
> Ancillary bus is intended to address below different use cases.
> 
> 1. ... single device different subsystem ... MFD not suitable platform not suitable description with audio example
>    Here cut of the 'may' part. Please definitive and you already have examples.
>    So right away say what purpose is served.
> 2. A PCI function support multiple class of devices such as RDMA, netdevice, and vDPA.
>    In this scenario, it is desired to maintain different class drivers for each of the class.
>    For such PCI device, for each class, an ancillary device with unique id should be created. Matching driver binds to >such ancillary device to create respective class device. This enables multiple class drivers to share same PCI function >and shared resources such as PCI BAR regions, irqs and more.
> 3. A PCI sub function is another type of device which represents a portion of the PCI device, similar to PCI virtual >function. Such device also needs matching service functionality. In this use case, for each PCI sub-function, an >ancillary device is created with unique id. A driver binds to such ancillary device to make use of a PCI sub-function >device. A PCI sub function based ancillary device will have one or more device attributes such as sub-function >number, NUMA node information, PCI BAR resources and more.
This is basically the same info that is already conveyed.  And the information about NUMA node and PCI BAR is outside the scope of the bus.  What information is shared after the initial probe of the ancillary_driver transparent to the bus.

Reply:
@paravmellanox I dont want to limit the usage here to only PCI devices. In the audio driver, we also use ancillary bus for ACPI platforms and we want to avoid limiting its usage by explicitly calling out PCI subfunctions. The motivation here was specifically chosen to be generic enough to appyly to all device types and not just PCI and I think we should keep it that way.

The implementation of the ancillary bus is being justified in this series with the audio driver example and we want to highlight that in the documentation so readers can connect with what is being said and what follows in the usage.

Reply:
@ranj063 for audio example, yes, point 1 covers it. So please keep solid audio example you have.

By mentioning the use-cases we are not limiting the usage of this bus by any means. Explicitly calling out the use case, we make sure that all stake holders are on-onboard of its intended use, even if other two use cases are on in this series.

Reply:
@ranj063 and I like to highlight as explained in great detail in PR 2265 that PCI BAR resources are not outside the scope of this bus. You might want to read the discussion in the closed PR 2265.

It is a generic resource that when device is added, it will be provided by the registering_driver.
ancilllary bus must have an API in future to support such addition (sfnum, resources and numa hierarchy details).
It is not part of this series, this is fine.
I will be extending ancillary bus incrementally to support this.
so yes, ancillary bus will be generic, just like PCI bus or for that matter any other bus, who will expose necessary attributes told by the registering driver.